### PR TITLE
Add safety checks (#681)

### PR DESCRIPTION
**Pull request by estyxx**
## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #681

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Etty for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/exceptions.py
Comment (BryceBeagle) [inline strawberry/exceptions.py:21]: This should just take the object itself, and not the `str` (i.e. the exception should be in charge of the output formatting)

File: tests/types/test_object_types.py
Comment (BryceBeagle) [inline tests/types/test_object_types.py:129]: This file (and directory of tests) is used for testing the `StrawberryType` logic of fields. We need to revisit the organization of tests in the future, but for now I would create a new test file in `tests/objects` for these

File: strawberry/exceptions.py
Comment (BryceBeagle) [inline strawberry/exceptions.py:17]: I think a better name for this exception might be something along the lines of `ObjectIsNotTypeError` (especially with the `Error` at the end)

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
I think this check is a really good idea!

File: strawberry/exceptions.py
Comment (estyxx) [reply strawberry/exceptions.py:21]: Mmm but if I give the `cls` object to the exception, this will be the non-class object, I need to pass the method the user is using: `type`|`input`|`interface`... 🤔

Comment (BryceBeagle) [reply strawberry/exceptions.py:21]: Oh right, I misread how this error was to be used. Hmmm, in that case I'm not a fan of passing the strings directly. Maybe we can use some `classmethod`s here?

```python
from __future__ import annotations
from enum import Enum

class NotAClass(Exception):
    class MethodType(Enum):
        INPUT = "input"
        INTERFACE = "interface"
        TYPE = "type"

    def __init__(self, obj: object, method_type: METHOD_TYPE):
        message = (
            f"strawberry.{method_type} can only be used with class types. Provided object "
            f"{obj} is not a type."
        )

        super().__init__(message)

   @classmethod
   def input(cls, obj: object) -> NotAClass:
       return cls(obj, cls.MethodType.INPUT)

   @classmethod
   def interface(cls) -> NotAClass:
       return cls(obj, cls.MethodType.INTERFACE)

   @classmethod
   def type(cls) -> NotAClass:
       return cls(obj, cls.MethodType.TYPE)
```
```python
a = object()  # example object that's not a type

raise NotAClass.interface(a)
```

Comment (estyxx) [reply strawberry/exceptions.py]: Nice! much better :)

File: strawberry/exceptions.py
Comment (BryceBeagle) [inline strawberry/exceptions.py:11]: This TODO still needs to be resolved

File: strawberry/object_type.py
Comment (BryceBeagle) [inline strawberry/object_type.py:158]: I think it would be cleaner to just use an `if/elif/else` tree

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
It also appears were missing some coverage in the tests -- but I'm not sure why. Some of the codecov alerts are for code that looks like it should be executed

File: strawberry/exceptions.py
Comment (estyxx) [reply strawberry/exceptions.py:11]: We should add a section in the docs that shows all these exceptions we can thrown?

**Comment by estyxx** [comment]
Think I figure out why the coverage dropped: pytest doesn't allow to have two files with the identical name in two different directories.

Also, could you please add the Oktoberfest-accepted label? tnks :)

File: strawberry/exceptions.py
Comment (BryceBeagle) [reply strawberry/exceptions.py:11]: Yes :+1: Can we open an issue and link to the issue number here in the comment?

Comment (estyxx) [reply strawberry/exceptions.py:11]: Here there is :) #1298

Comment (BryceBeagle) [reply strawberry/exceptions.py:11]: Excellent. In the case that I was ambiguous, I meant in the source comment too :slightly_smiling_face:

File: tests/objects/test_types.py
Comment (BryceBeagle) [inline tests/objects/test_types.py:10]: It doesn't end up changing things here, but you probably wanted to do this to each of these expected errors:
```suggestion
        r"strawberry.type can only be used with class types. Provided "
        r"object .* is not a type."
```

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

**Review by BryceBeagle — APPROVED** [review]

File: tests/objects/test_types.py
Comment (estyxx) [reply tests/objects/test_types.py]: Yeah the match is working so think python is converting them to be a rstring both... but agree, it's also looks better in the IDE :)

Comment (BryceBeagle) [reply tests/objects/test_types.py]: It's working because the second string has no special characters. Note that `r""` doesn't make a regex string, it makes a _raw_ string, where escape characters (i.e. those prefixed by `\`s) are ignored and interpreted literally. However, people often use `r""` strings to indicate regex patterns even if not using special characters

Comment (estyxx) [reply tests/objects/test_types.py]: Oh, you are right, I can omit now the `r`... Was not working at some point, so I added the `r`... I gues there was something else wrong :D

Comment (BryceBeagle) [reply tests/objects/test_types.py]: I tend to use the `r""` strings for regex even if there's no escape sequences. I think it adds a little expressiveness to the code saying "hey, this is a pattern used as a regex". Could help future maintainers. But I'll leave the choice to use it or not up to you (or @patrick91 if he has an opinion :P)

Comment (estyxx) [reply tests/objects/test_types.py]: Yeah totally agree! I will leave it there :)

File: tests/objects/test_types.py
Comment (patrick91) [inline tests/objects/test_types.py:37]:
```suggestion

    with pytest.raises(ObjectIsNotClassError, match=expected_error):
```

File: tests/objects/test_types.py
Comment (patrick91) [inline tests/objects/test_types.py:26]:
```suggestion

    with pytest.raises(ObjectIsNotClassError, match=expected_error):
        @strawberry.input
```

File: tests/objects/test_types.py
Comment (patrick91) [inline tests/objects/test_types.py:14]:
```suggestion

    with pytest.raises(ObjectIsNotClassError, match=expected_error):
        @strawberry.type
```

**Review by patrick91 — APPROVED** [review]
Nice one! I've only requested a couple of nits, will commit them now :)

**Pull request merged**
